<a href="https://colab.research.google.com/github/Sweta-Das/TensorFlow-Python-Projects/blob/Fundamentals/2-Linear_Regression/optimization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Optimization
- It is the process of adjusting the parameters (weights and biases) of a model to minimize (or maximize) a certain objective function (like loss function).
- Goal of optimization is to minimize the loss so that the model becomes more accurate.

## Importance of Optimization in Training Linear Regression Model
- Optimization is crucial for training the linear regression model as the model's accuracy directly depends on finding the best values of $w$ and $b$.
- Poor optimization can lead to underfitting or overfitting.

### Steps to Optimize Linear Regression Model
1. **Initializing parameters** -> Assigning random values for $w$ and $b$
2. **Forward Pass** -> Computing the predictions $ŷ$
3. **Calculate loss** -> Using a loss function to measure the error between $ŷ$ and $y$
4. **Backpropagation** -> Computing the gradients of the loss function w.r.t $w$ and $b$
5. **Update Parameters** -> Adjusting $w$ and $b$ using an optimization algo. like *Gradient Descent*
6. The process is repeated until the loss converges to a minimum or reaches a stopping criterion.

In [1]:
# Importing libraries
import numpy as np
import tensorflow as tf

In [12]:
# Generating synthetic data for training & testing
np.random.seed(42)
X_train = np.random.rand(100, 1) # 100 samples, 1 feature
y_train = 4 * X_train + np.random.randn(100, 1) # Linear relation with some noise

X_test = np.random.rand(20, 1)
y_test = 4 * X_test + np.random.randn(20, 1)

In [13]:
# Creating linear regression model in tensorflow
class LinearRegressionModel(tf.Module):
  def __init__(self):
    # Initializing weight and bias randomly
    self.w = tf.Variable(tf.random.normal([1]), name="weight")
    self.b = tf.Variable(tf.random.normal([1]), name="bias")

  def __call__(self, X):
    return self.w * X + self.b

model = LinearRegressionModel()

In [15]:
# Loss Function (Mean Squared Error)
def loss_fn(y_true, y_pred):
  return tf.reduce_mean(tf.square(y_true - y_pred))

In [16]:
# Training
def train(model, X_train, y_train, epochs):
  # Initializing optimizer after model creation
  optimizer = tf.optimizers.SGD(learning_rate=0.1)

  for epoch in range(epochs):
    with tf.GradientTape() as tape:
      predictions = model(X_train)
      loss = loss_fn(y_train, predictions)

    # Computing gradients and updating parameters
    gradients = tape.gradient(loss, [model.w, model.b])
    optimizer.apply_gradients(zip(gradients, [model.w, model.b]))

    if epoch % 10 == 0:
      print(f"Epoch {epoch}: Loss = {loss.numpy():.4f}, w = {model.w.numpy()}, b = {model.b.numpy()}")

train(model, X_train, y_train, epochs=100)

Epoch 0: Loss = 4.5440, w = [0.25911975], b = [0.5524367]
Epoch 10: Loss = 1.3363, w = [1.1261972], b = [1.3596457]
Epoch 20: Loss = 1.1972, w = [1.4720178], b = [1.2571218]
Epoch 30: Loss = 1.1003, w = [1.7482758], b = [1.121551]
Epoch 40: Loss = 1.0274, w = [1.9864544], b = [1.0012794]
Epoch 50: Loss = 0.9726, w = [2.1929066], b = [0.89683014]
Epoch 60: Loss = 0.9314, w = [2.3719232], b = [0.80624974]
Epoch 70: Loss = 0.9005, w = [2.527154], b = [0.72770417]
Epoch 80: Loss = 0.8772, w = [2.6617596], b = [0.65959466]
Epoch 90: Loss = 0.8597, w = [2.77848], b = [0.6005348]


In [17]:
# Testing

# Defining accuracy function (MAE and RMSE)
def mean_abs_error(y_true, y_pred):
  return tf.reduce_mean(tf.abs(y_true-y_pred))

def root_mean_sqr_error(y_true, y_pred):
  return tf.sqrt(tf.reduce_mean(tf.square(y_true - y_pred)))

def test(model, X_test, y_test):
  predictions = model(X_test)
  loss = loss_fn(y_test, predictions)
  mae = mean_abs_error(y_test, predictions)
  rmse = root_mean_sqr_error(y_test, predictions)
  print(f"Test MAE: {mae.numpy(): .4f}, Test RMSE: {rmse.numpy():.4f}, Test Loss: {loss.numpy():.4f}")
  return mae, rmse

In [19]:
test(model=model, X_test=X_test, y_test=y_test)

Test MAE:  0.6896, Test RMSE: 0.8188, Test Loss: 0.6704


(<tf.Tensor: shape=(), dtype=float32, numpy=0.6896304>,
 <tf.Tensor: shape=(), dtype=float32, numpy=0.8187523>)

 - MAE = 0.6896 means that, on average, the model is relatively close to the actual test values but still has room for improvement.
 - RMSE = 0.8188 means that there are some larger errors in predictions, though the errors are relatively moderate.
 - Test loss = 0.6704 is consistent with the RMSE, suggesting a reasonable fit, though improvements could be made.